packages/gatsby-source-contentful/src/source-nodes.js

```diff
@@ -335,14 +366,20 @@ export async function sourceNodes(
       }
 
       // Remove references to deleted nodes
-      if (n.sys.id && deletedEntryGatsbyReferenceIds.size) {
-        Object.keys(n).forEach(name => {
-          // @todo Detect reference fields based on schema. Should be easier to achieve in the upcoming version.
-          if (!name.endsWith(`___NODE`)) {
-            return
-          }
-          if (Array.isArray(n[name])) {
-            n[name] = n[name].filter(referenceId => {
+      if (
+        n.sys.id &&
+        deletedEntryGatsbyReferenceIds.size &&
+        referenceFieldMap.has(n.sys.contentType)
+      ) {
+        const referenceFields = [
+          ...referenceFieldMap.get(n.sys.contentType),
+          ...reverseReferenceFields,
+        ]
```

Comment:
This field name merge could happen earlier, but will be rewritten anyways very soon due to planned structural changes in the (automatic) reverse links